File: sdks/python/src/delopay/http.py
from __future__ import annotations

import json
import time
from typing import Any
from urllib.error import HTTPError, URLError
from urllib.parse import urlencode, urljoin
from urllib.request import Request, urlopen

from .errors import ApiError

IDEMPOTENT_METHODS = {"GET", "HEAD"}


class HttpClient:
    def __init__(
        self,
        *,
        api_key: str,
        base_url: str,
        timeout_ms: int,
        max_retries: int,
    ) -> None:
        if not api_key:
            raise ValueError("api_key is required")

        self._api_key = api_key
        self._base_url = base_url
        self._timeout_seconds = timeout_ms / 1000
        self._max_retries = max_retries

    def request(
        self,
        method: str,
        path: str,
        payload: dict[str, Any] | None = None,
        query: dict[str, Any] | None = None,
    ) -> dict[str, Any] | None:
        method_upper = method.upper()
        retries = self._max_retries if method_upper in IDEMPOTENT_METHODS else 0

        for attempt in range(retries + 1):
            try:
                return self._send_once(method_upper, path, payload, query)
            except HTTPError as exc:
                body_text = exc.read().decode("utf-8") if exc.fp else ""
                parsed = _parse_json(body_text)
                request_id = exc.headers.get("x-request-id") if exc.headers else None
                code = None
                message = exc.msg or "Request failed"

                if isinstance(parsed, dict):
                    message = str(parsed.get("message") or parsed.get("error") or message)
                    code = parsed.get("code") or parsed.get("errorCode")
                    request_id = request_id or parsed.get("requestId")

                if exc.code >= 500 and attempt < retries:
                    _sleep(attempt)
                    continue

                raise ApiError(
                    status=exc.code,
                    message=message,
                    code=str(code) if code is not None else None,
                    request_id=str(request_id) if request_id is not None else None,
                    raw=parsed if parsed is not None else body_text,
                ) from exc
            except URLError as exc:
                if attempt < retries:
                    _sleep(attempt)
                    continue

                raise ApiError(status=0, message="Network request failed", raw=str(exc.reason)) from exc

        raise ApiError(status=0, message="Request exhausted retries")

    def _send_once(
        self,
        method: str,
        path: str,
        payload: dict[str, Any] | None,
        query: dict[str, Any] | None,
    ) -> dict[str, Any] | None:
        url = self._build_url(path, query)
        data = json.dumps(payload).encode("utf-8") if payload is not None else None

        headers = {
            "Authorization": f"Bearer {self._api_key}",
            "Accept": "application/json",
        }
        if data is not None:
            headers["Content-Type"] = "application/json"

        request = Request(url=url, data=data, method=method, headers=headers)
        with urlopen(request, timeout=self._timeout_seconds) as response:
            raw = response.read().decode("utf-8")
            if not raw:
                return None
            return json.loads(raw)

    def _build_url(self, path: str, query: dict[str, Any] | None) -> str:
        base = self._base_url if self._base_url.endswith("/") else f"{self._base_url}/"
        clean_path = path[1:] if path.startswith("/") else path
        url = urljoin(base, clean_path)

        if not query:
            return url

        filtered = {key: value for key, value in query.items() if value is not None}
        if not filtered:
            return url

        return f"{url}?{urlencode(filtered)}"


def _parse_json(raw: str) -> Any:
    if not raw:
        return None

    try:
        return json.loads(raw)
    except json.JSONDecodeError:
        return raw


def _sleep(attempt: int) -> None:
    delay = min(1.0, 0.1 * (2**attempt))
    time.sleep(delay)

File: sdks/typescript/src/http.ts
import { ApiError } from "./errors";

type QueryValue = string | number | boolean | undefined;

type RequestOptions = {
  method: "GET" | "POST" | "PUT" | "DELETE" | "PATCH" | "HEAD";
  path: string;
  body?: unknown;
  query?: Record<string, QueryValue>;
};

type HttpClientConfig = {
  apiKey: string;
  baseUrl: string;
  timeoutMs: number;
  maxRetries: number;
};

const IDEMPOTENT_METHODS = new Set(["GET", "HEAD"]);

export class HttpClient {
  private readonly config: HttpClientConfig;

  constructor(config: HttpClientConfig) {
    this.config = config;
  }

  async request<T>(options: RequestOptions): Promise<T> {
    const url = this.buildUrl(options.path, options.query);
    const retries = IDEMPOTENT_METHODS.has(options.method) ? this.config.maxRetries : 0;

    for (let attempt = 0; attempt <= retries; attempt += 1) {
      try {
        const response = await this.runFetch(url, options);
        if (response.ok) {
          if (response.status === 204) {
            return undefined as T;
          }

          const text = await response.text();
          return text ? (JSON.parse(text) as T) : (undefined as T);
        }

        const rawText = await response.text();
        const parsed = parseJson(rawText);
        const status = response.status;
        const requestId =
          response.headers.get("x-request-id") ??
          (isObject(parsed) && typeof parsed.requestId === "string" ? parsed.requestId : undefined);
        const code = extractCode(parsed);
        const message = extractMessage(parsed) ?? response.statusText ?? "Request failed";

        if (status >= 500 && attempt < retries) {
          await sleep(backoffMs(attempt));
          continue;
        }

        throw new ApiError({
          status,
          message,
          code,
          requestId,
          raw: parsed ?? rawText
        });
      } catch (error) {
        if (error instanceof ApiError) {
          throw error;
        }

        if (attempt < retries && isRetryableTransportError(error)) {
          await sleep(backoffMs(attempt));
          continue;
        }

        throw new ApiError({
          status: 0,
          message: "Network request failed",
          raw: error
        });
      }
    }

    throw new ApiError({
      status: 0,
      message: "Request exhausted retries"
    });
  }

  private async runFetch(url: string, options: RequestOptions): Promise<Response> {
    const controller = new AbortController();
    const timeout = setTimeout(() => controller.abort(), this.config.timeoutMs);

    try {
      const headers: Record<string, string> = {
        Authorization: `Bearer ${this.config.apiKey}`,
        Accept: "application/json"
      };

      if (options.body !== undefined) {
        headers["Content-Type"] = "application/json";
      }

      return await fetch(url, {
        method: options.method,
        headers,
        body: options.body !== undefined ? JSON.stringify(options.body) : undefined,
        signal: controller.signal
      });
    } finally {
      clearTimeout(timeout);
    }
  }

  private buildUrl(path: string, query?: Record<string, QueryValue>): string {
    const url = new URL(path, ensureTrailingSlash(this.config.baseUrl));

    if (query) {
      for (const [key, value] of Object.entries(query)) {
        if (value !== undefined) {
          url.searchParams.set(key, String(value));
        }
      }
    }

    return url.toString();
  }
}

function parseJson(raw: string): unknown {
  if (!raw) {
    return undefined;
  }

  try {
    return JSON.parse(raw);
  } catch {
    return raw;
  }
}

function extractMessage(raw: unknown): string | undefined {
  if (!isObject(raw)) {
    return undefined;
  }

  if (typeof raw.message === "string") {
    return raw.message;
  }

  if (typeof raw.error === "string") {
    return raw.error;
  }

  return undefined;
}

function extractCode(raw: unknown): string | undefined {
  if (!isObject(raw)) {
    return undefined;
  }

  if (typeof raw.code === "string") {
    return raw.code;
  }

  if (typeof raw.errorCode === "string") {
    return raw.errorCode;
  }

  return undefined;
}

function isObject(input: unknown): input is Record<string, any> {
  return typeof input === "object" && input !== null;
}

function isRetryableTransportError(error: unknown): boolean {
  if (error instanceof TypeError) {
    return true;
  }

  if (typeof error === "object" && error !== null && "name" in error) {
    return (error as { name?: string }).name === "AbortError";
  }

  return false;
}

function backoffMs(attempt: number): number {
  return Math.min(1000, 100 * 2 ** attempt);
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

function ensureTrailingSlash(input: string): string {
  return input.endsWith("/") ? input : `${input}/`;
}
